File: packages/gg-sandbox-rust/src/skills.rs
//! Read the skills this run authored.
//!
//! The names available are listed in the system prompt, and an unknown one comes back as `NotFound`
//! carrying the full list.

use crate::bindings::test_cabinet::gg::skills;
use crate::core::ApiError;
use crate::wire;

/// The gg tools this module dispatches — see [`files::OPERATIONS`](crate::files::OPERATIONS).
pub(crate) const OPERATIONS: &[&str] = &["read_skill"];

/// Read a skill by name and hand back its body with the front matter stripped.
///
/// The body is pinned permanently into context.
///
/// A skill may be code rather than prose, or as well as it. Its code becomes a crate every later
/// program this session writes reaches as `<key>::<name>`, and using it opens a documentation view
/// of each function that crate declares. An on-use program runs once this program has ended, and
/// whatever it shows arrives on the next turn.
///
/// # Arguments
///
/// * `name` — The skill's name, as the system prompt lists it.
///
/// # Returns
///
/// The skill's body with its front matter stripped.
///
/// # Errors
///
/// `NotFound` — listing the skills that do exist — when the name is unknown.
#[doc(alias = "ggop:skills.read_skill")]
pub fn read_skill(name: &str) -> Result<String, ApiError> {
    wire::lift(skills::read_skill(name))
}

File: packages/gg-sandbox/src/gg/skills.ts
/**
 * Read the skills available to this session.
 *
 * A skill name is a plain string. The names on offer are listed in the system prompt, and an unknown
 * one comes back as `not-found` carrying the list of the ones that exist.
 */

import * as raw from "test-cabinet:gg/skills";
import { call } from "../internal/errors.js";

/**
 * Read a skill by name, returning its body with the front matter stripped.
 *
 * The body is pinned permanently into the context window.
 *
 * A skill may carry code as well as, or instead of, prose. Reading one makes its module importable
 * by every later program and opens a documentation view of each function the module declares, which
 * states the line that imports it. Nothing about the module is added to this reply. An on-use script
 * runs after the program has ended, on every read, and whatever it shows arrives on the next turn.
 *
 * @ggop skills.read_skill
 * @param name The skill's name, as the system prompt lists it.
 * @returns the skill's body, with its front matter stripped.
 * @throws `ApiError` with `not-found` — listing the skills that do exist — when the name is
 * unknown.
 */
export function readSkill(name: string): string {
  return call(() => raw.readSkill(name));
}
